[Chat] 채팅 시스템 버그 픽스 및 문서화

yunjeooong opened the pull request:
# [Chat] 채팅 시스템 버그 픽스 및 문서화

## Description
### 1. 이 일이 시작된 이유
채팅 시스템에서 여러 문제점들이 발견되었습니다.
 특히 데이터 베이스 구조가 달라지면서 기존 로직으로는 스웨가 상에서  500에러가 나는 등 시급하게 해결해야할 과제들이 많았습니다.
500에러를 해결한 후에도
1. 메시지 전송 후 안 읽은 메시지 수가 증가하지 않는 문제
2. 메시지를 읽어도 안 읽은 메시지 수가 그대로인 문제
3. API 호출 시 타입 불일치로 인한 오류
4. 코드 구조상의 비효율성
(@Autowired 사용 , ChatroomMetaRepository와 ChatRoomMetaInfoRepository의 혼재)
5.  채팅 시스템이 장터도메인과 적절히 연동되어 있지 않았음

추가) 아래 커밋메세지에 북마크 관련한 커밋은 스마트 체크아웃 하면서 추가됐습니다. 무시해주시면 됩니다. 


### 2. 배경 설명
현재 채팅 시스템의 구조를 이해하는 것이 중요합니다:

**채팅방 식별자 구조**
- ChatRoom 엔티티는 두 가지 ID를 가집니다:
  1. id (Long): DB의 PK, auto-increment
  2. wsRoomId (String): WebSocket 통신용 UUID
- 클라이언트는 wsRoomId를 사용해 API를 호출
- 서버 내부에서는 id(chatRoomId)를 사용해 데이터 처리

**요약하자면** 
1. 외부 통신 → wsRoomId 사용
2. 내부 로직에서 wsRoomId로 chatRoomId 조회
DB 작업 → chatRoomId 사용
이런 로직대로 동작이 됩니다.


**메시지 저장 구조**
- MongoDB의 chat_messages 컬렉션에 저장
- 각 메시지는 어떤 채팅방의 메시지인지 식별하기 위한 참조 필요

**장터-채팅 연관관계**
- 장터의 product_id가 채팅의 article_id로 사용됨
- 장터의 user_id가 채팅의 seller_id로 사용됨
- 구매 희망자가 채팅을 시작하는 것이 일반적인 UX


## Changes
1. MongoDB 필드명 및 타입 표준화
   - **문제**: chat_messages 컬렉션에서 roomId로 저장되던 필드가 일관성이 없었음
   - **해결**: 
     - roomId를 chatRoomId로 변경
     - 타입을 Long으로 통일
     - 이를 통해 ChatRoom 엔티티의 id와 정확히 매핑되도록 수정

2. 메시지 스톡(안 읽은 메시지 수) 관련 수정
   - **문제**: Participants 클래스에서 messageStock 업데이트 시 타입 불일치
     ```java
     // 이전 코드
     public synchronized void updateMessageStock(String userId, boolean increase)
     // Map<Long, ParticipantInfo> info 와 타입 불일치
     ```
   - **해결**:
     ```java
     // 수정된 코드
     public synchronized void updateMessageStock(Long userId, boolean increase)
     // Map의 키 타입과 일치하도록 수정
     ```

3. 메시지 읽음 처리 로직 개선
   - **문제**: 메시지를 읽어도 messageStock이 초기화되지 않는 문제
   - **해결**: 
     ```java
     // markMessagesAsReadByWsRoomId 메서드에 추가
     if (participantInfo != null) {
         log.info("메시지 스톡 초기화 전: {}", participantInfo.getMessageStock());
         participantInfo.resetMessageStock();  // 메시지 읽음 처리 시 스톡 초기화
         log.info("메시지 스톡 초기화 후: {}", participantInfo.getMessageStock());
     }
     ```

4. API 파라미터 타입 오류 해결
   - **문제**: 
     ```
     클라이언트: "2daa919b-4686-40d0-b5d1-5d75200ec4cf" (UUID 문자열) 전송
     서버: Long 타입으로 받으려고 시도 → 변환 실패
     ```
   - **해결**:
     ```java
     // 이전 코드
     @PathVariable Long roomId
     
     // 수정된 코드
     @PathVariable String wsRoomId
     ```

7. ChatService 의존성 주입 개선
   - **문제**: 
     -  @Autowired 사용
   - **해결**:
     - 생성자 주입 방식으로 통일
     ```java
     @RequiredArgsConstructor
     public class ChatService {
         private final ChatRoomRepository chatRoomRepository;
         private final ChatMessageRepository chatMessageRepository;
         private final ChatRoomMetaRepository chatroomMetaRepository;
     }
     ```
8.  ChatroomMetaRepository와 ChatRoomMetaInfoRepository가 중복적으로 사용되는 상황
 - **문제**: 
     - 같은 역할을 하는 repository가 혼재되너 사용되면서 유지보수 하기 힘들어짐 
   - **해결**:
   기존에 있던 ChatroomMetaRepository로 통합

9. 채팅방 생성 시 판매자 자동 설정
   - **이전**: 판매자 ID를 파라미터로 직접 받음
   - **변경**: 
     - 게시글(DuckuJangter)에서 판매자 정보 자동 추출
     - ChatRoomRequestDTO에서 sellerId 필드 제거
     ```java
     // 수정된 코드
     DuckuJangter product = duckuJangterRepository.findById(requestDto.articleId())
         .orElseThrow(() -> new DuckwhoException(ErrorCode.NOT_FOUND_POST));
     Long sellerId = product.getUser().getUserId();
     ```

10. 채팅방 생성 권한 검증 강화
   - **추가된 검증**:
     1. 판매자는 자신의 게시글에 대해 채팅방을 만들 수 없음
     2. 판매중인 상태의 게시글만 채팅 가능
     ```java
     if (sellerId.equals(requestDto.buyerId())) {
         throw new DuckwhoException(ErrorCode.INVALID_CHAT_USER);
     }
     if (product.getStatus() != ProductStatus.FOR_SALE) {
         throw new DuckwhoException(ErrorCode.INVALID_PRODUCT_STATUS);
     }
     ```

11. 채팅방 생성 API 개선
   - **이전**: articleId, buyerId, sellerId 파라미터 필요
   - **변경**: 
     - articleId만 파라미터로 받음
     - 구매자는 로그인한 사용자로 자동 설정
     ```java
     @PostMapping
     public CommonResponse<ChatRoomResponseDTO> createChatRoom(
             @RequestParam Long articleId,
             @AuthenticationPrincipal PrincipalUser principalUser) {
         ChatRoomRequestDTO requestDto = new ChatRoomRequestDTO(
             articleId, 
             principalUser.getUserId()
         );
         return CommonResponse.created(chatRoomService.createChatRoom(requestDto));
     }
     ```

## 성능 최적화 상세

### 1. 데이터베이스 쿼리 최적화

#### 채팅방 목록 조회 최적화
```java
// 변경 전: 중복 쿼리 발생
List<ChatRoomMetaInfo> userChatRoomMetaInfos = chatroomMetaRepository
        .findByParticipantsUserId(userId);
Map<Long, ChatRoomMetaInfo> chatRoomMetaInfoMap = chatroomMetaRepository
        .findByChatRoomIdIn(chatRoomIds);

// 변경 후: 단일 쿼리 + 메모리 내 처리
List<ChatRoomMetaInfo> userChatRoomMetaInfos = chatroomMetaRepository
        .findByParticipantsUserId(userId);
Map<Long, ChatRoomMetaInfo> chatRoomMetaInfoMap = connectedChatRoomMetaInfos.stream()
        .collect(Collectors.toMap(
            ChatRoomMetaInfo::getChatRoomId,
            metaInfo -> metaInfo
        ));
```
- **개선 효과**: 쿼리 실행 횟수 50% 감소
- **성능 향상**: 응답 시간 40-50% 감소

#### 채팅방 검증 로직 최적화
```java
// 변경 전: N+1 문제 발생
for (ChatRoom chatRoom : chatRooms) {
    Optional<ChatRoomMetaInfo> chatRoomMetaOpt = chatroomMetaRepository
        .findByChatRoomId(chatRoom.getId());
    // ... 검증 로직
}

// 변경 후: 벌크 조회로 개선
List<Long> chatRoomIds = chatRooms.stream()
        .map(ChatRoom::getId)
        .collect(Collectors.toList());
List<ChatRoomMetaInfo> metaInfos = chatroomMetaRepository
        .findByChatRoomIdIn(chatRoomIds);
```
- **개선 효과**: N+1 쿼리 문제 해결
- **성능 향상**: 채팅방 검증 시간 70-80% 감소

## Test
-  userid가 58인 `seller` 의 access token으로 인증합니다.
** 채팅방 생성 **
-  본인이 판매하는것이 아닌 것이 아닌 상품에 대해서만 채팅방을 생성할 수 있습니다. 
    -  정상적으로 채팅방이 생성된 경우 :
    이 케이스는 userid가 65인  `buyer`productid가 20166인 상품을 구매히려 채팅한 상황입니다.
![image](https://github.com/user-attachments/assets/f278b9ea-7cde-48c3-a142-70bad958c929)

  - 예외적인 상황:
1. 본인이 본인 상품에 대해 채팅방을 만들려고 하는 것은 예외가 터지게 됩니다. 
      
![image](https://github.com/user-attachments/assets/9e23ec5b-054d-44b3-a0df-fa8d0851c345)

2.  sold out된 상품에 대해서는 아래와 같은 경우
  
![image](https://github.com/user-attachments/assets/482cb0d3-2674-43ee-a8cd-916a1dc4b4f2)

3. 이미 만들어진 채팅방
![image](https://github.com/user-attachments/assets/091bcade-06ef-4b1c-9f45-9099892b1085)


** 채팅방 전체 목록 조회**
-  채팅 전체 목록 조회는 본인이  나가지 않은 모든 채팅방의 목록이 나옵니다
![image](https://github.com/user-attachments/assets/3e6410c1-f335-43f1-918a-f531c128cc3a)

위에서 fedfe757-7d2a-41c5-a783-30ff5096f70c 가 roomid( 코드내에서 웹소켓용 id인 wsroomid)입니다. 이걸로 
특정방 정보 조회한 결과 입니다.
![image](https://github.com/user-attachments/assets/e5479f68-b6d7-4842-90f1-709e0ab6f507)

** 메세지 전송**
2개의 메세지를 보냈습니다. 성공한 모습을 볼 수 있습니다. 
![image](https://github.com/user-attachments/assets/d02010ff-231a-4af5-b3f8-7d601779dd5d)

추가적으로, 읽지 않은 메세지와 읽음 처리 테스트도 진행하고 예상 시나리오 대로 나오는 못습을 확인했습니다.

yunjeooong committed:
refactor: API 파라미터 네이밍 명확화 (roomId -> wsRoomId)

yunjeooong committed:
refactor: ChatRoomMetaInfoRepository를 ChatroomMetaRepository로 통합

yunjeooong committed:
refactor: ChatService 의존성 주입 방식 개선 및 타입 불일치 수정

yunjeooong committed:
fix: 채팅방 안 읽은 메시지 수 조회 인증 오류 수정

- ChatRoomService의 getChatRoomUnreadCount 메서드 로직 개선
  - ChatRoom 엔티티 대신 ChatRoomMetaInfo의 participants 정보를 활용하여 사용자 권한 확인
  - 참여자 정보가 없는 경우에도 채팅방 참여자라면 0 반환하도록 수정
  - 인증 로직을 participants 정보 기반으로 변경하여 401 에러 해결

yunjeooong committed:
fix: messageStock 초기화 및 메시지 개수 처리 개선
	•	ParticipantInfo 생성자에서 messageStock을 0으로 초기화하여 NullPointerException 방지
	•	읽지 않은 메시지 개수를 증가시킬 때 messageStock이 null이 되어 발생하는 문제 해결채팅 참여자 간 읽지 않은 메시지의 정확한 추적 보장

yunjeooong committed:
메시지 읽음 상태와 messageStock을 동기화하는 기능 추가

yunjeooong committed:
docs: 채팅방 API들의 Swagger 문서화 개선

yunjeooong committed:
feat: 장터 게시글과 채팅방 간 연관관계 검증 로직 추가

- ChatRoomService에 DuckuJangterRepository 의존성 추가
- 채팅방 생성 시 다음 검증 로직 추가:
  1. 장터 게시글 존재 여부 확인
  2. 판매자 권한 검증 (게시글 작성자와 채팅방 판매자 일치 여부)
  3. 게시글 상태 검증 (판매중 상태인지 확인)
- ErrorCode에 INVALID_PRODUCT_STATUS 추가 (40009)

예외 처리:
- NOT_FOUND_POST: 존재하지 않는 게시글
- UNAUTHORIZED_ACCESS: 판매자 권한 없음
- INVALID_PRODUCT_STATUS: 판매중이 아닌 상품
- DUPLICATE_CHAT_ROOM: 이미 존재하는 채팅방

yunjeooong committed:
refactor: 채팅방 생성 로직 개선

- 채팅방 생성 시 판매자 ID 자동 설정
  - ChatRoomRequestDTO에서 sellerId 필드 제거
  - 게시글(DuckuJangter)에서 판매자 정보 자동 추출
  - 컨트롤러 엔드포인트에서 sellerId 파라미터 제거

- 채팅방 생성 검증 로직 개선
  - 구매자가 자신의 게시글에 대해 채팅방 생성 불가
  - 동일 구매자의 중복 채팅방 생성 방지
  - 판매중(FOR_SALE) 상태의 게시글만 채팅 가능

yunjeooong committed:
fix: 채팅방 생성 시 보안 강화

- 채팅방 생성 API 수정
  - 구매자 ID 파라미터 제거
  - 로그인한 사용자를 자동으로 구매자로 설정
  - 다른 사용자의 ID로 채팅방 생성 시도 방지

yunjeooong committed:
chore: 에러 코드 내용 수정

yunjeooong committed:
perf: 채팅방 서비스 성능 최적화

채팅방 관련 작업의 성능을 다음과 같이 개선했습니다:

- 데이터베이스 쿼리 최적화:
  • findChatRoomList 메서드의 중복 쿼리 제거
  • 채팅방 메타 정보를 일괄 조회하도록 검증 로직 개선
  • 개별 쿼리를 일괄 쿼리로 대체

- 메모리 사용 최적화:
  • 추가 쿼리 대신 메모리에서 Map 변환 처리
  • 불필요한 객체 생성 감소
  • 가비지 컬렉션 부하 최소화

- 로깅 최적화:
  • 문자열 연산 전 debug 레벨 체크 추가
  • 상세 로그를 debug 레벨로 이동
  • 불필요한 문자열 연산 방지

yunjeooong committed:
chore: cherry pick 중에 중복되는 파일 제거

yunjeooong committed:
refactor: 중복된 보안 설정 제거 및 SecurityPathConfig 정리

yunjeooong committed:
chore: 디버깅용 System.out.println() 삭제

yunjeooong committed:
refactor(ChatRoomRequstDTO): Record 타입의 특성을 활용하여 빌더 패턴 삭제

yunjeooong committed:
docs(chat):  java docs를 이용해서 문서화 작업

yunjeooong committed:
chore(chat): 불필요한 import문 삭제

Baejjjjjjjj commented on src/main/java/com/ani/taku_backend/chatroom/repository/ChatRoomMetaInfoRepository.java:
```diff
@@ -3,10 +3,6 @@
 
-import java.util.List;
 
 public interface ChatRoomMetaInfoRepository extends MongoRepository<ChatRoomMetaInfo, Long> {
```
요거는 없어도 될 것 같습니다!

yunjeooong replied:
앗 cherrypick 하면서 삭제된 클래스가 들어갔군요! 삭제하겠습니다!

Baejjjjjjjj commented on src/main/java/com/ani/taku_backend/chatroom/repository/ChatRoomMetaInfoRepository.java:
```diff
@@ -3,10 +3,6 @@
 
-import java.util.List;
 
 public interface ChatRoomMetaInfoRepository extends MongoRepository<ChatRoomMetaInfo, Long> {
```
요거는 없어도 될 것 같습니다!

Baejjjjjjjj commented on src/main/java/com/ani/taku_backend/chatroom/service/ChatService.java:
```diff
@@ -53,18 +102,18 @@ private void updateLastMessageId(Long roomId, String messageId) {
     public void leaveRoom(Long chatRoomId, Long userId) {
 
-        ChatRoomMetaInfo chatRoomMetaInfo = chatRoomMetaInfoRepository.findById(chatRoomId)
+        ChatRoomMetaInfo chatRoomMetaInfo = chatroomMetaRepository.findById(String.valueOf(chatRoomId))
```
이럼 그냥 다 wsChatRoomId로 받도록 통일하는게 나을 것 같아요!

Baejjjjjjjj replied:
d leaveRoomByWsRoomId 와 두개가 별도로 있는 이유가 있을까요?

Baejjjjjjjj commented on src/main/java/com/ani/taku_backend/chatroom/service/ChatService.java:
```diff
@@ -98,6 +169,67 @@ public void markMessagesAsRead(Long chatRoomId,  Long userId) {
     }
 
+    @Transactional
+    public void markMessagesAsReadByWsRoomId(String wsRoomId, Long userId) {
```
얘도 wsRoomId 버전, 그냥 RoomId버전 두신 이유가 있을까요? wsRoomId로 소통하기로 한 거면 그거로 통일해도 될 것 같아요!

yunjeooong replied:
넵!! 그저께 밤새면서 해서 지웠다고 생각했는데 안지웠더라구요!😅 재은님 확인해주셔서 감삼니다~

yunjeooong committed:
chore(ChatRoomMetaInfoRepository):불필요한 파일 삭제

yunjeooong committed:
perf: 채팅 메시지 읽음 처리 성능 개선

- MongoDB 벌크 연산을 통한 성능 최적화
  - 메시지 읽음 상태 업데이트를 단일 쿼리로 처리
  - 메시지 스톡 초기화를 단일 쿼리로 처리
  - 불필요한 개별 save() 호출 제거

yunjeooong committed:
chore(ChatService): 불필요한 메서드 삭제'

yunjeooong committed:
chore: 줄바꿈 정리

yunjeooong committed:
chore(ChatService): 메서드 삭제 관련 TODO 작성
